influxdb/metric.go

```diff
@@ -138,82 +185,101 @@ func (m *Metric) buildQuery(startTime, endTime time.Time) client.Query {
 func (m *Metric) metricDescriptor() *metricpb.MetricDescriptor {
 	return &metricpb.MetricDescriptor{
 		Type:        m.StackdriverName(),
-		MetricKind:  metricpb.MetricDescriptor_GAUGE,
+		MetricKind:  m.metricKind(),
 		ValueType:   metricpb.MetricDescriptor_DOUBLE,
 		Description: fmt.Sprintf("InfluxDB query: %s", m.Name),
 		DisplayName: m.config.Query,
 	}
 }
 
-func (m *Metric) filterPoints(points []point, endTime time.Time) ([]point, error) {
-	if !m.config.TimeAggregated {
+// metricKind returns Stackdriver metric kind for this metric.
+func (m *Metric) metricKind() metricpb.MetricDescriptor_MetricKind {
+	if m.config.Cumulative {
+		return metricpb.MetricDescriptor_CUMULATIVE
+	}
+	return metricpb.MetricDescriptor_GAUGE
+}
+
+func (m *Metric) filterPoints(lastPoint, endTime time.Time, points []point) ([]point, error) {
+	if !m.config.TimeAggregated && !m.config.Cumulative {
 		// For non-time-aggregated gauge metrics, we don't need to apply any
 		// filtering since we only query InfluxDB for new points, and points
 		// that are too fresh are ignored by applying an offset.
 		return points, nil
 	}
 
-	queryInterval, err := m.config.queryInterval()
-	if err != nil {
-		return nil, err
-	}
-
 	var filteredPoints []point
 	for _, p := range points {
-		// Once cumulative metrics come in play, we'll filter using lastPoint.
-		// For now we can rely on our constructed query to only fetch the
-		// latest points.
-
-		// For time-aggregated InfluxQL queries with interval i, Influx returns
-		// points with timestamp x, denoting the aggregated value from time
-		// [i, i + x).
-		intervalEnd := p.timestamp.Add(queryInterval)
-
-		// If this time interval hasn't finished accumulating data, wait for
-		// it to complete next time around.
-		if intervalEnd.UnixNano() > endTime.UnixNano() {
+		if m.config.TimeAggregated {
+			// For time-aggregated InfluxQL queries with interval i, Influx
+			// returns points with timestamp x, denoting the aggregated value
+			// from time [i, i + x). Hence, i + x is used to denote its
+			// end time.
+			queryInterval, err := m.config.queryInterval()
+			if err != nil {
+				return nil, err
+			}
+
+			p.timestamp = p.timestamp.Add(queryInterval)
+		}
+
+		// If this (time-aggregated) time interval hasn't finished accumulating
+		// data, wait for it to complete next time around.
+		if p.timestamp.UnixNano() > endTime.UnixNano() {
```

Comment:
Is this necessary for cumulative metrics that are not time-aggregated? If not, perhaps this should be moved to the `if m.config.TimeAggregated` block above?

Reply:
Moved!